core/trino-main/src/test/java/io/trino/execution/TestQueryStateMachine.java

```diff
@@ -410,6 +418,72 @@ public void testUpdateMemoryUsage()
         assertThat(stateMachine.getPeakTaskRevocableMemory()).isEqualTo(10);
     }
 
+    @Test
+    public void testPreserveFirstFailure()
+            throws Exception
+    {
+        CountDownLatch cleanup = new CountDownLatch(1);
+        QueryStateMachine queryStateMachine = queryStateMachine()
+                .withMetadata(new TracingMetadata(noopTracer(), createTestMetadataManager())
+                {
+                    @Override
+                    public void cleanupQuery(Session session)
+                    {
+                        cleanup.countDown();
+                        super.cleanupQuery(session);
+                    }
+                })
+                .build();
+
+        Future<?> anotherThread = executor.submit(() -> {
+            awaitUninterruptibly(cleanup, 10, SECONDS);
+            queryStateMachine.transitionToFailed(new IllegalStateException("Second exception"));
+        });
+        Future<?> failingThread = executor.submit(() -> {
+            queryStateMachine.transitionToFailed(new TrinoException(TYPE_MISMATCH, "First exception"));
+        });
+
+        failingThread.get(10, SECONDS);
+        anotherThread.get(10, SECONDS);
+
+        ExecutionFailureInfo failureInfo = queryStateMachine.getFinalQueryInfo().orElseThrow().getFailureInfo();
+        assertThat(failureInfo).isNotNull();
+        assertThat(failureInfo.getErrorCode()).isEqualTo(TYPE_MISMATCH.toErrorCode());
+        assertThat(failureInfo.getMessage()).isEqualTo("First exception");
+    }
+
+    @Test
+    public void testPreserveCancellation()
+            throws Exception
+    {
+        CountDownLatch cleanup = new CountDownLatch(1);
+        QueryStateMachine queryStateMachine = queryStateMachine()
+                .withMetadata(new TracingMetadata(noopTracer(), createTestMetadataManager())
+                {
+                    @Override
+                    public void cleanupQuery(Session session)
+                    {
+                        cleanup.countDown();
+                        super.cleanupQuery(session);
+                    }
+                })
+                .build();
+
+        Future<?> anotherThread = executor.submit(() -> {
+            awaitUninterruptibly(cleanup, 10, SECONDS);
+            queryStateMachine.transitionToFailed(new IllegalStateException("Second exception"));
+        });
+        Future<?> cancellingThread = executor.submit(queryStateMachine::transitionToCanceled);
+
+        cancellingThread.get(10, SECONDS);
+        anotherThread.get(10, SECONDS);
+
+        // TODO queryStateMachine.getFinalQueryInfo() does not exist for cancelled queries, but may be created by anotherThread due to a race
```

Comment:
For some reason transitionToFailed and transitionToCanceled have different code paths and only first sets `finalQueryInfo`. that's probably a miss in 8a89c29bff8303a23ee6633605cb0e521acaf578?
cc @JamesRTaylor

Reply:
fixed in another commit

Reply:
I wonder if https://github.com/trinodb/trino/pull/20231 is related

Reply:
good question. don't know